mithril-common/src/test_utils/certificate_chain_builder.rs

```diff
@@ -0,0 +1,752 @@
+                let next_fixture = fixtures_per_epoch.get(&epoch.next()).unwrap_or_else(|| panic!("Next fixture not found at epoch {epoch:?} with {} total certificates and {} certificates per epoch", self.total_certificates, self.certificates_per_epoch));
+                let context = CertificateChainBuilderContext {
+                    index_certificate,
+                    total_certificates,
+                    epoch,
+                    fixture,
+                    next_fixture,
+                };
+                match index_certificate {
+                    0 => genesis_certificate_processor(
+                        self.build_genesis_certificate(&context, &genesis_signer),
+                        &context,
+                        &genesis_signer,
+                    ),
+                    _ => standard_certificate_processor(
+                        self.build_standard_certificate(&context),
+                        &context,
+                    ),
+                }
+            })
+            .collect::<Vec<Certificate>>();
+        let certificates_chained = self.compute_chained_certificates(certificates);
+
+        (certificates_chained, genesis_verifier)
+    }
+
+    fn compute_clerk_for_signers(signers: &[SignerFixture]) -> ProtocolClerk {
+        let first_signer = &signers[0].protocol_signer;
+
+        ProtocolClerk::from_signer(first_signer)
+    }
+
+    fn compute_avk_for_signers(signers: &[SignerFixture]) -> ProtocolAggregateVerificationKey {
+        let clerk = Self::compute_clerk_for_signers(signers);
+
+        clerk.compute_avk().into()
+    }
+
+    fn setup_genesis() -> (ProtocolGenesisSigner, ProtocolGenesisVerifier) {
+        let genesis_signer = ProtocolGenesisSigner::create_deterministic_genesis_signer();
+        let genesis_verifier = genesis_signer.create_genesis_verifier();
+
+        (genesis_signer, genesis_verifier)
+    }
+
+    fn build_epochs_sequence(&self) -> impl Iterator<Item = Epoch> {
+        let total_certificates = self.total_certificates;
+        let certificates_per_epoch = self.certificates_per_epoch;
+        assert!(
+            certificates_per_epoch > 0,
+            "Certificates per epoch must be greater than 0"
+        );
+        assert!(
+            total_certificates >= certificates_per_epoch,
+            "Total certificates must be greater or equal to certificates per epoch"
+        );
+        // The total number of epochs in the sequence is the total number of certificates divided by
+        // the number of certificates per epoch plus two (one for the genesis epoch and one to compute the next fixtures of the last epoch)
+        let total_epochs_in_sequence =
+            (total_certificates - 1).div_ceil(certificates_per_epoch) + 2;
```

Comment:
I understand the `+2` part, but why `(total_certificates - 1)`?

Reply:
This is because the genesis epoch will have only one certificate.
I enhanced the readability of the formula and I gave some more details in the explanation comment :+1: